File: sat.py
# 5
import pandas as pd 
import numpy as np  
# Importing the dataset
spa = pd.read_csv("C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Linear Regression\\SAT_GPA.csv")
# Performing EDA
spa.describe()
# Graphical Representation
import matplotlib.pyplot as plt 
# Barplot
plt.bar(height = spa.SAT_Scores, x = np.arange(1, 201, 1))
# Histogram
plt.hist(spa.SAT_Scores) 
# Boxplot
plt.boxplot(spa.SAT_Scores) 
plt.bar(height = spa.GPA, x = np.arange(1, 201, 1))
plt.hist(spa.GPA) #histogram
plt.boxplot(spa.GPA) #boxplot
# Scatter plot
plt.scatter(x = spa['GPA'], y = spa['SAT_Scores'], color = 'green') 
# correlation
np.corrcoef(spa.GPA, spa.SAT_Scores) 
# Covariance
cov_output = np.cov(spa.GPA, spa.SAT_Scores)[0, 1]
cov_output
spa.cov()
import statsmodels.formula.api as smf
# Simple Linear Regression
model = smf.ols('SAT_Scores ~ GPA', data = spa).fit()
model.summary()
pred1 = model.predict(pd.DataFrame(spa['GPA']))
# Regression Line
plt.scatter(spa.GPA, spa.SAT_Scores)
plt.plot(spa.GPA, pred1, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res1 = spa.SAT_Scores - pred1
res_sqr1 = res1 * res1
mse1 = np.mean(res_sqr1)
rmse1 = np.sqrt(mse1)
rmse1
# Model building on Transformed Data
# Log Transformation
# x = log(GPA); y = SAT_Scores
plt.scatter(x = np.log(spa['GPA']), y = spa['SAT_Scores'], color = 'brown')
np.corrcoef(np.log(spa.GPA), spa.SAT_Scores) #correlation
model2 = smf.ols('SAT_Scores ~ np.log(GPA)', data = spa).fit()
model2.summary()
pred2 = model2.predict(pd.DataFrame(spa['GPA']))
# Regression Line
plt.scatter(np.log(spa.GPA), spa.SAT_Scores)
plt.plot(np.log(spa.GPA), pred2, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res2 = spa.SAT_Scores - pred2
res_sqr2 = res2 * res2
mse2 = np.mean(res_sqr2)
rmse2 = np.sqrt(mse2)
rmse2
# Exponential transformation
# x = GPA; y = log(SAT_Scores)
plt.scatter(x = spa['GPA'], y = np.log(spa['SAT_Scores']), color = 'orange')
np.corrcoef(spa.GPA, np.log(spa.SAT_Scores)) #correlation
model3 = smf.ols('np.log(SAT_Scores) ~ GPA', data = spa).fit()
model3.summary()
pred3 = model3.predict(pd.DataFrame(spa['GPA']))
pred3_at = np.exp(pred3)
pred3_at
# Regression Line
plt.scatter(spa.GPA, np.log(spa.SAT_Scores))
plt.plot(spa.GPA, pred3, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res3 = spa.SAT_Scores - pred3_at
res_sqr3 = res3 * res3
mse3 = np.mean(res_sqr3)
rmse3 = np.sqrt(mse3)
rmse3
# Polynomial transformation
# x = GPA; x^2 = GPA*GPA; y = log(SAT_Scores)
model4 = smf.ols('np.log(SAT_Scores) ~ GPA + I(GPA*GPA)', data = spa).fit()
model4.summary()
pred4 = model4.predict(pd.DataFrame(spa))
pred4_at = np.exp(pred4)
pred4_at
# Regression line
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 2)
X = spa.iloc[:, 1:2].values
X_poly = poly_reg.fit_transform(X)
y = spa.iloc[:, 1].values
plt.scatter(spa.GPA, np.log(spa.SAT_Scores))
plt.plot(X, pred4, color = 'red')
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res4 = spa.SAT_Scores - pred4_at
res_sqr4 = res4 * res4
mse4 = np.mean(res_sqr4)
rmse4 = np.sqrt(mse4)
rmse4
# Choose the best model using RMSE
data = {"MODEL":pd.Series(["SLR", "Log model", "Exp model", "Poly model"]), "RMSE":pd.Series([rmse1, rmse2, rmse3, rmse4])}
table_rmse = pd.DataFrame(data)
table_rmse
# The best model
from sklearn.model_selection import train_test_split
train, test = train_test_split(spa, test_size = 0.3)
finalmodel = smf.ols('np.log(SAT_Scores) ~ GPA + I(GPA*GPA)',  data= train).fit()
finalmodel.summary()
# Predict on test dSalarya
test_pred = finalmodel.predict(pd.DataFrame(test))
pred_test_Salary = np.exp(test_pred)
pred_test_Salary
# Model Evaluation on Test data
test_res = test.SAT_Scores - pred_test_Salary
test_sqrs = test_res * test_res
test_mse = np.mean(test_sqrs)
test_rmse = np.sqrt(test_mse)
test_rmse
# Prediction on train data
train_pred = finalmodel.predict(pd.DataFrame(train))
pred_train_AT = np.exp(train_pred)
pred_train_AT
# Model Evaluation on train data
train_res = train.SAT_Scores - pred_train_AT
train_sqrs = train_res * train_res
train_mse = np.mean(train_sqrs)
train_rmse = np.sqrt(train_mse)
train_rmse

File: sat.R
# 5
library(readr)
# Importing the data
st.at <- read.csv('C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Linear Regression\\SAT_GPA.csv', header = T)
View(st.at)
# Performing EDA
summary(st.at)
library(Hmisc)
describe(st.at)
library(lattice) 
# Graphical exploration
# Dotplot
dotplot(st.at$SAT_Scores, main = "Dot Plot of Waist Circumferences")
dotplot(st.at$GPA, main = "Dot Plot of Adipose Tissue Areas")
# Boxplot
boxplot(st.at$SAT_Scores, col = "dodgerblue4")
boxplot(st.at$GPA, col = "red", horizontal = T)
# Histogram
hist(st.at$SAT_Scores)
hist(st.at$GPA)
# Normal QQ plot
qqnorm(st.at$SAT_Scores)
qqline(st.at$SAT_Scores)
qqnorm(st.at$GPA)
qqline(st.at$GPA)
hist(st.at$SAT_Scores, prob = TRUE)          # prob=TRUE for probabilities not counts
lines(density(st.at$SAT_Scores))             # add a density estimate with defaults
lines(density(st.at$SAT_Scores, adjust = 2), lty = "dotted")   # add another "smoother" density
hist(st.at$GPA, prob = TRUE)          # prob=TRUE for probabilities not counts
lines(density(st.at$GPA))             # add a density estimate with defaults
lines(density(st.at$GPA, adjust = 2), lty = "dotted")   # add another "smoother" density
# Bi-variate analysis
# Scatter plot
plot(st.at$SAT_Scores,st.at$GPA, main = "Scatter Plot", col = "Dodgerblue4", col.main = "Dodgerblue4", col.lab = "Dodgerblue4", xlab = "Salary_hike", ylab = "Churn_out_rate", pch = 20)  
# Correlation Coefficient
cor(st.at$SAT_Scores,st.at$GPA)
# Co-variance
cov(st.at$SAT_Scores,st.at$GPA)
# Linear Regression model
reg <- lm(st.at$SAT_Scores~ st.at$GPA, data = st.at) 
summary(reg)
confint(reg, level = 0.95)
pred <- predict(reg, interval = "predict")
pred <- as.data.frame(pred)
View(pred)
# ggplot for adding Regression line for data
library(ggplot2)
ggplot(data = st.at, aes(st.at$GPA, st.at$GPA) ) + geom_point() + stat_smooth(method = lm, formula = y ~ x)
# Evaluation the model for fitness 
cor(pred$fit,  st.at$GPA)
reg$residuals
rmse <- sqrt(mean(reg$residuals^2))
rmse
# Transformation Techniques
# input = log(x); output = y
plot(log(st.at$SAT_Scores), st.at$GPA)
cor(log(st.at$SAT_Scores), st.at$GPA)
reg_log <- lm(st.at$GPA ~ log(st.at$SAT_Scores), data = st.at)
summary(reg_log)
confint(reg_log,level = 0.95)
pred <- predict(reg_log, interval = "predict")
pred <- as.data.frame(pred)
cor(pred$fit, st.at$GPA)
rmse <- sqrt(mean(reg_log$residuals^2))
rmse
# Regression line for data
ggplot(data = st.at, aes(log(st.at$SAT_Scores), st.at$GPA) ) + geom_point() + stat_smooth(method = lm, formula = y ~ log(x))
# Log transformation applied on 'y'
# input = x; output = log(y)
plot(st.at$SAT_Scores, log(st.at$GPA))
cor(st.at$SAT_Scores, log(st.at$GPA))
reg_log1 <- lm(log(st.at$GPA) ~ st.at$SAT_Scores, data = st.at)
summary(reg_log1)
predlog <- predict(reg_log1, interval = "predict")
predlog <- as.data.frame(predlog)
reg_log1$residuals
sqrt(mean(reg_log1$residuals^2))
# Exponential function
pred <- exp(predlog)   
pred <- as.data.frame(pred)
cor(pred$fit, st.at$GPA)
res_log1 = st.at$GPA - pred$fit
rmse <- sqrt(mean(res_log1^2))
rmse
# Regression line for data
ggplot(data = st.at, aes( st.at$SAT_Scores, log(st.at$GPA )) ) + geom_point() + stat_smooth(method = lm, formula = log(y) ~ x)
# Non-linear models = Polynomial models
# input = x & x^2 (2-degree) and output = log(y)
reg2 <- lm(log(st.at$GPA ) ~ st.at$SAT_Scores + I(st.at$SAT_Scores*st.at$SAT_Scores), data = st.at)
summary(reg2)
predlog <- predict(reg2, interval = "predict")
pred <- exp(predlog)
pred <- as.data.frame(pred)
cor(pred$fit,st.at$GPA)
res2 =st.at$GPA - pred$fit
rmse <- sqrt(mean(res2^2))
rmse
# Regression line for data
ggplot(data = st.at, aes(st.at$SAT_Scores, log(st.at$GPA)) ) + geom_point() + stat_smooth(method = lm, formula = y ~ poly(x, 2, raw = TRUE))
# Data Partition

# Random Sampling
n <- nrow(st.at)
n1 <- n * 0.8
n2 <- n - n1

train_ind <- sample(1:n, n1)
train <- st.at[train_ind, ]
test <-  st.at[-train_ind, ]

# Non-random sampling
train <- st.at[1:100, ]
test <- st.at[101:200, ]

log(train$GPA)
log(test$GPA)
plot(train$GPA, log(train$SAT_Scores))
plot(test$GPA, log(test$SAT_Scores))
model <- lm(log(st.at$GPA) ~ st.at$SAT_Scores + I(st.at$SAT_Scores * st.at$SAT_Scores), data = train)
summary(model)
confint(model,level=0.95)
log_res <- predict(model,interval = "confidence", newdata = test)
predict_original <- exp(log_res) # converting log values to original values
predict_original <- as.data.frame(predict_original)
test_error <- test$GPA- predict_original$fit # calculate error/residual
test_error
test_rmse <- sqrt(mean(test_error^2))
test_rmse
log_res_train <- predict(model, interval = "confidence", newdata = train)
predict_original_train <- exp(log_res_train) # converting log values to original values
predict_original_train <- as.data.frame(predict_original_train)
train_error <- train$GPA - predict_original_train$fit # calculate error/residual
train_error
train_rmse <- sqrt(mean(train_error^2))
train_rmse

